coordinator.go

```diff
@@ -2,69 +2,252 @@ package influxdb
+		shardMapping.MapPoint(shardInfo, p)
+	}
+	return shardMapping, nil
 }
 
 // Write is coordinates multiple writes across local and remote data nodes
 // according the request consistency level
 func (c *Coordinator) Write(p *WritePointsRequest) error {
+	shardMappings, err := c.MapShards(p)
+	if err != nil {
+		return err
+	}
 
-	// FIXME: use the consistency level specified by the WritePointsRequest
-	pol := newConsistencyPolicyN(1)
+	// Write each shard in it's own goroutine and return as soon
+	// as one fails.
+	ch := make(chan error, len(shardMappings.Points))
+	for shardID, points := range shardMappings.Points {
+		go func(shard meta.ShardInfo, points []data.Point) {
+			ch <- c.writeToShards(shard, p.ConsistencyLevel, points)
+		}(shardMappings.Shards[shardID], points)
+	}
 
-	// FIXME: build set of local and remote point writers
-	ws := []PointsWriter{}
+	for range shardMappings.Points {
+		select {
+		case <-c.closing:
+			return ErrWriteFailed
+		case err := <-ch:
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// writeToShards writes points to a shard and ensures a write consistency level has been met.  If the write
+// partially succceds, ErrPartialWrite is returned.
+func (c *Coordinator) writeToShards(shard meta.ShardInfo, consistency ConsistencyLevel, points []data.Point) error {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	// Abort early if there are no writer configured yet
+	if len(c.shardWriters) == 0 {
+		return ErrWriteFailed
+	}
 
+	// The required number of writes to achieve the requested consistency level
+	required := len(shard.OwnerIDs)
+	switch consistency {
+	case ConsistencyLevelAny, ConsistencyLevelOne:
+		required = 1
+	case ConsistencyLevelQuorum:
+		required = required/2 + 1
+	}
+
+	// holds the response to the ShardWriter.Write calls
 	type result struct {
-		writerID int
-		err      error
-	}
-	ch := make(chan result, len(ws))
-	for i, w := range ws {
-		go func(id int, w PointsWriter) {
-			err := w.Write(p)
-			ch <- result{id, err}
-		}(i, w)
-	}
-	timeout := time.After(defaultReadTimeout)
-	for range ws {
+		wrote int
+		err   error
+	}
+
+	// response channel for each shard writer go routine
+	ch := make(chan result, len(c.shardWriters))
+
+	for _, w := range c.shardWriters {
+		// write to each ShardWriter (local and remote), in parallel
+		go func(w ShardWriter, shardID uint64, points []data.Point) {
+			wrote, err := w.WriteShard(shardID, points)
+			ch <- result{wrote, err}
+		}(w, shard.ID, points)
+	}
+
+	var wrote int
+	timeout := time.After(defaultWriteTimeout)
+	for range c.shardWriters {
 		select {
+		case <-c.closing:
+			return ErrWriteFailed
 		case <-timeout:
 			// return timeout error to caller
 			return ErrTimeout
 		case res := <-ch:
-			if !pol.IsDone(res.writerID, res.err) {
-				continue
-			}
-			if res.err != nil {
+			wrote += res.wrote
+
+			// ErrShardNotLocal might be returned from a local writer. Ignore it.
```

Comment:
Why are local writers created for shards that aren't local? Seems odd to be doing this and returning an error that we're always ignoring.